File: android/app/src/main/java/com/compartirarchivosred/app/net/Framed.kt
package com.compartirarchivosred.app.net

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

/**
 * Enmarcado: líneas JSON delimitadas por '\n' intercaladas con bloques de bytes
 * crudos de tamaño conocido. Un búfer interno evita "comerse" bytes de payload.
 */
class Framed(private val input: InputStream, private val output: OutputStream) {
    private val buf = ByteArray(8192)
    private var start = 0
    private var end = 0

    fun readLine(): String? {
        val sb = ArrayList<Byte>(128)
        while (true) {
            if (start >= end) {
                end = input.read(buf, 0, buf.size)
                start = 0
                if (end <= 0) return if (sb.isEmpty()) null else String(sb.toByteArray(), Charsets.UTF_8)
            }
            val b = buf[start++]
            if (b == '\n'.code.toByte()) return String(sb.toByteArray(), Charsets.UTF_8)
            if (b != '\r'.code.toByte()) sb.add(b)
        }
    }

    fun copyExact(dest: OutputStream, count: Long, onProgress: (Long) -> Unit) {
        var remaining = count
        while (remaining > 0) {
            if (start >= end) {
                end = input.read(buf, 0, buf.size)
                start = 0
                if (end <= 0) throw IOException("Conexión cerrada durante la recepción.")
            }
            val avail = end - start
            val take = minOf(avail.toLong(), remaining).toInt()
            dest.write(buf, start, take)
            start += take
            remaining -= take
            onProgress(count - remaining)
        }
    }

    fun writeLine(s: String) {
        output.write((s + "\n").toByteArray(Charsets.UTF_8))
        output.flush()
    }

    fun writeBytes(src: InputStream, count: Long, onProgress: (Long) -> Unit) {
        val tmp = ByteArray(81920)
        var sent = 0L
        while (sent < count) {
            val toRead = minOf(tmp.size.toLong(), count - sent).toInt()
            val n = src.read(tmp, 0, toRead)
            if (n <= 0) break
            output.write(tmp, 0, n)
            sent += n
            onProgress(sent)
        }
        output.flush()
    }
}

File: windows/src/App/Net/Framed.cs
using System.IO;
using System.Text;

namespace CompartirArchivosRED.Net;

/// <summary>
/// Lector/escritor con enmarcado: líneas JSON delimitadas por '\n' intercaladas
/// con bloques de bytes crudos de tamaño conocido. Un búfer interno evita
/// "comerse" bytes de payload al leer una línea.
/// </summary>
public sealed class Framed
{
    private readonly Stream _s;
    private readonly byte[] _buf = new byte[8192];
    private int _start;
    private int _end;

    public Framed(Stream s) => _s = s;

    public string? ReadLine()
    {
        var sb = new List<byte>(128);
        while (true)
        {
            if (_start >= _end)
            {
                _end = _s.Read(_buf, 0, _buf.Length);
                _start = 0;
                if (_end <= 0)
                    return sb.Count == 0 ? null : Encoding.UTF8.GetString(sb.ToArray());
            }
            byte b = _buf[_start++];
            if (b == (byte)'\n') return Encoding.UTF8.GetString(sb.ToArray());
            if (b != (byte)'\r') sb.Add(b);
        }
    }

    public void CopyExact(Stream dest, long count, Action<long> onProgress)
    {
        long remaining = count;
        while (remaining > 0)
        {
            if (_start >= _end)
            {
                _end = _s.Read(_buf, 0, _buf.Length);
                _start = 0;
                if (_end <= 0) throw new IOException("Conexión cerrada durante la recepción.");
            }
            int avail = _end - _start;
            int take = (int)Math.Min(avail, remaining);
            dest.Write(_buf, _start, take);
            _start += take;
            remaining -= take;
            onProgress(count - remaining);
        }
    }

    public void WriteLine(string s)
    {
        byte[] data = Encoding.UTF8.GetBytes(s + "\n");
        _s.Write(data, 0, data.Length);
        _s.Flush();
    }

    public void WriteBytes(Stream src, long count, Action<long> onProgress)
    {
        byte[] tmp = new byte[81920];
        long sent = 0;
        while (sent < count)
        {
            int toRead = (int)Math.Min(tmp.Length, count - sent);
            int n = src.Read(tmp, 0, toRead);
            if (n <= 0) break;
            _s.Write(tmp, 0, n);
            sent += n;
            onProgress(sent);
        }
        _s.Flush();
    }
}
